hid-uclogic-core.c

```diff
@@ -285,6 +289,39 @@ static int uclogic_raw_event(struct hid_device *hdev,
 		}
 	}
 
+	/* Tweak frame control reports, if necessary */
+	if ((report->type == HID_INPUT_REPORT) &&
+	    (report->id == params->frame_id)) {
+		/* If need to, and can, set pad device ID for Wacom drivers */
+		if (params->frame_dev_id_byte > 0 &&
+		    params->frame_dev_id_byte < size) {
+			data[params->frame_dev_id_byte] = 0xf;
+		}
+		/* If need to, and can, read rotary encoder state change */
+		if (params->frame_re_lsb > 0 &&
+		    params->frame_re_lsb / 8 < size) {
+			unsigned byte = params->frame_re_lsb / 8;
+			unsigned bit = params->frame_re_lsb % 8;
+			/* Read state */
+			u8 state = (data[byte] >> bit) & 0x3;
+			/* Encode change into 2-bit signed integer */
+			u8 change = 0;
+			if (state == 0 || state == 3) {
+				u8 xor = drvdata->re_state ^ state;
+				if (xor == 1) {
+					change = 1;
+				} else if (xor == 2) {
+					change = 3;
+				}
```

Comment:
First of all, the wheel on the tablet reports rotary encoder states in Gray coding, that is (in binary): 00, 01, 11, 10 (i.e. 0, 1, 3, 2). The wheel clicks as it rotates and two of the states are reported in the stable positions (00, 11) and two in between them (01, 10). We're supposed to only report a single change of rotation on stable positions (00 and 11). Furthermore, we can detect which way the wheel is rotating be checking for the previous state.

With just two bit Gray coding it's easy to describe all possible combinations. I don't remember which way would be which, but it doesn't really matter. Given:

    10->00 CW
    01->11 CW
    01->00 CCW
    10->11 CCW

we can write the following statements:

    if ((prev_state == 2 && state = 0) ||
        (prev_state == 1 && state = 3)) {
        /* Clockwise */
    } else if ((prev_state == 1 && state == 0) ||
               (prev_state == 2 && state == 3)) {
        /* Counter-clockwise */
    }

In the code I'm trying to be a smart-ass and am using XOR to avoid checking the previous state, but since I have to check for stable states anyway, and I have just 2-bit Gray coding, I end up with the same number of conditions and add XOR to that. So I might as well rewrite it to be simpler.

However, perhaps you're asking about `change` being 1 or 3. That is simply unsigned representation of 1 and -1 2-bit two's complement signed integers, which will be decoded by the generic HID driver according to the report descriptor I wrote.

Reply:
The tablet firmware engineers should have just done the decoding themselves.